proposals/exception-handling/Exceptions.md

```diff
@@ -205,99 +193,166 @@ flushed, the embedder defines how to handle uncaught exceptions. Otherwise, the
 found enclosing try block is the catching try block.
 
 A throw inside the body of a catch block is never caught by the corresponding
-try block of the catch block, since instructions in the body of the catch block
-are not in the body of the try block.
+try block of the catch/unwind block, since instructions in the body of the
+catch/unwind block are not in the body of the try block.
 
 Once a catching try block is found for the thrown exception, the operand stack
-is popped back to the size the operand stack had when the try block was entered,
-and then an `exnref` referring to the caught exception is pushed back onto the
-operand stack.
+is popped back to the size the operand stack had when the try block was entered.
+
+Then in case of a try-catch block, tagged catch blocks are tried in the
+order they appear in the catching try block, until one matches. If a matched
+tagged catch block is found, control is transferred to the body of the catch
+block, and the data fields of the exception are pushed back onto the stack.
+
+Otherwise, control is transferred to the body of the `catch_all` block, if any.
+However, unlike tagged catch blocks, the constructor arguments are not copied
+back onto the operand stack. If no tagged catch blocks were matched, and the
+catching try block doesn't have a `catch_all` block, the exception is rethrown
+to the next enclosing try block.
 
 If control is transferred to the body of a catch block, and the last instruction
 in the body is executed, control then exits the try block.
 
-If the selected catch block does not throw an exception, it must yield the
-value(s) expected by the corresponding catching try block. This includes popping
-the caught exception.
+In case of a try-unwind block, the control is transfered to the body of the
+`unwind` block. After executing instructions within the `unwind` block, the
+exception is automatically rethrown when the control reaches the `end`
+instruction. As in the case of the `catch_all` block, the exception arguments
+are not copied onto the operand stack.
+
+If the selected catch/unwind block does not throw an exception, it must yield
+the value(s) expected by the corresponding catching try block. This includes
+popping the caught exception.
 
 Note that a caught exception can be rethrown using the `rethrow` instruction.
 
 ### Rethrowing an exception
 
-The `rethrow` instruction takes the exception associated with the `exnref` on
-top of the stack, and rethrows the exception. A rethrow has the same effect as a
-throw, other than an exception is not created. Rather, the referenced exception
-on top of the stack is popped and then thrown. The `rethrow` instruction traps
-if the value on the top of the stack is null.
+The `rethrow` instruction can only appear in the body of a catch/catch_all block
+but not within the body of an unwind block. It always re-throws the exception
```

Comment:
I think saying that `rethrow` cannot appear in an `unwind` block is too restrictive, consider the following example:
```wasm
try
...
catch
  try
  ...
  unwind
    rethrow 1
  end
end
```
The `rethrow` appears in the `unwind` block, but IIUC this is ok because we never rethrow its exception. Should the rule rather be something like "the `rethrow` immediate can only be a `catch` or `catch_all` block"?

Reply:
I still don't understand why we are making this restriction. IIUC, for `N > 0` the following code

```
try bt
  instr1*
unwind
  instr2*
  rethrow N
  instr3*
end
```

is equivalent to

```
try bt
  instr1*
unwind
  instr2*
  br 0
  instr3*
end
rethrow N-1
```

For N=0, 

```
try bt
  instr1*
unwind
  instr2*
  rethrow 0
  instr3*
end
```

is equivalent to 


```
try bt
  instr1*
unwind
  instr2*
end
```

In both cases, `instr3*` is redundant. 

So I don't understand what situation is prevented with this restriction. Am I (still) misunderstanding the semantics of `unwind`?

(cc. @RossTate)

Reply:
Clarification: I do agree with @thibaudmichaud that, if we do add a restriction, then it should be on the rethrow target (i.e., that rethrow cannot target an unwind block).

Reply:
Agreed with @thibaudmichaud and @ioannad. The restriction is stated the wrong way. It should simply say that the rethrow target must be a catch block, which automatically implies that it has to be properly nested into one.

I also agree that there is no particular reason to disallow rethrow targetting unwind.

Reply:
Forbidding rethrow would make sense for finally, because that has nothing to rethrow if the block was entered regularly, but no real point for unwind (unless one assumes we will be adding yet some other form of abort that is neither a trap nor an exception, but that's hardly desirable). But it's conservative to disallow it, so okay to leave it out for now.

Reply:
@rossberg IIUC also having `br`, `br_if`, `br_table`, or `return` among the instructions of an `unwind` clause would abort the unwinding process.

Reply:
A `rethrow` that "targets" `unwind` is problematic because `unwind` can fire due to things besides exceptions. But a `rethrow` that is simply a `throw` of a caught exception should be allowed within `unwind`, just like `throw` is. (For supporting Common Lisp, it is necessary that exceptions be throwable from within `unwind` clauses.)

Reply:
@ioannad, but branches do not invoke intermediate unwind clauses, IIUC.

Reply:
I clarified the rethrow-unwind relationship to match the rule we agreed on here: https://github.com/WebAssembly/exception-handling/pull/137/commits/4d6c635ce79413fd9f48abe87050d1926575a986